File: examples/python/part-7/23-incident-response-recovery.py
# Illustrative examples for notes/part-7-testing-compliance/23-incident-response-recovery.md
# Not for production use. Licensed under MIT (see LICENSE-CODE).

from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Dict, List


class Severity(str, Enum):
    CRITICAL = "Critical"
    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"


class IncidentStatus(str, Enum):
    OPEN = "open"
    CONTAINED = "contained"
    RECOVERING = "recovering"
    RESOLVED = "resolved"


@dataclass
class Incident:
    id: str
    title: str
    severity: Severity
    status: IncidentStatus
    run_ids: List[str] = field(default_factory=list)
    affected_tools: List[str] = field(default_factory=list)
    affected_users: List[str] = field(default_factory=list)
    data_types: List[str] = field(default_factory=list)
    detected_at: datetime | None = None
    owner: str = ""
    summary: str = ""


@dataclass
class Signal:
    event: str
    tool: str = ""
    run_id: str = ""
    attrs: Dict[str, str] = field(default_factory=dict)


def classify(signal: Signal) -> Severity:
    if signal.event in (
        "secret_exfiltrated",
        "cross_tenant_leakage",
        "production_shell_executed",
    ):
        return Severity.CRITICAL
    if signal.event in (
        "egress_with_secret_blocked",
        "mcp_server_compromised",
        "unsafe_tool_executed",
    ):
        return Severity.HIGH
    if signal.event in (
        "prompt_injection_detected",
        "tool_denied_repeated",
        "budget_exceeded",
    ):
        return Severity.MEDIUM
    return Severity.LOW


class ContainmentAction(str, Enum):
    STOP_RUN = "stop_run"
    DISABLE_TOOL = "disable_tool"
    BLOCK_EGRESS = "block_egress"
    ROTATE_SECRET = "rotate_secret"
    ISOLATE_MEMORY = "isolate_memory"
    READ_ONLY_MODE = "read_only_mode"
    DISABLE_MCP = "disable_mcp_server"


def build_containment_plan(signal: Signal) -> List[ContainmentAction]:
    if signal.event == "secret_exfiltrated":
        return [
            ContainmentAction.STOP_RUN,
            ContainmentAction.BLOCK_EGRESS,
            ContainmentAction.ROTATE_SECRET,
            ContainmentAction.DISABLE_TOOL,
        ]
    if signal.event == "mcp_server_compromised":
        return [
            ContainmentAction.DISABLE_MCP,
            ContainmentAction.BLOCK_EGRESS,
            ContainmentAction.ISOLATE_MEMORY,
        ]
    if signal.event == "memory_poisoning_detected":
        return [ContainmentAction.STOP_RUN, ContainmentAction.ISOLATE_MEMORY]
    if signal.event == "unsafe_tool_executed":
        return [
            ContainmentAction.STOP_RUN,
            ContainmentAction.DISABLE_TOOL,
            ContainmentAction.READ_ONLY_MODE,
        ]
    return [ContainmentAction.STOP_RUN]


def validate_incident(incident: Incident) -> None:
    if not incident.id or not incident.title or not incident.owner:
        raise ValueError("incident has required empty fields")

    if not incident.severity or not incident.status:
        raise ValueError("incident has no severity or status")

    if not incident.run_ids:
        raise ValueError("incident has no run ids")

    if incident.detected_at is None:
        raise ValueError("incident has no detected_at")


def export_incident(incident: Incident) -> str:
    validate_incident(incident)
    payload = {
        "id": incident.id,
        "title": incident.title,
        "severity": incident.severity.value,
        "status": incident.status.value,
        "run_ids": incident.run_ids,
        "affected_tools": incident.affected_tools,
        "affected_users": incident.affected_users,
        "data_types": incident.data_types,
        "detected_at": incident.detected_at.isoformat() if incident.detected_at else None,
        "owner": incident.owner,
        "summary": incident.summary,
    }
    return json.dumps(payload, indent=2)

File: examples/typescript/part-7/23-incident-response-recovery.ts
// Illustrative examples for notes/part-7-testing-compliance/23-incident-response-recovery.md
// Not for production use. Licensed under MIT (see LICENSE-CODE).

enum Severity {
  Critical = "Critical",
  High = "High",
  Medium = "Medium",
  Low = "Low",
}

enum IncidentStatus {
  Open = "open",
  Contained = "contained",
  Recovering = "recovering",
  Resolved = "resolved",
}

interface Incident {
  id: string;
  title: string;
  severity: Severity;
  status: IncidentStatus;
  runIds: string[];
  affectedTools: string[];
  affectedUsers: string[];
  dataTypes: string[];
  detectedAt: Date;
  owner: string;
  summary: string;
}

interface Signal {
  event: string;
  tool: string;
  runId: string;
  attrs: Record<string, string>;
}

function classify(signal: Signal): Severity {
  switch (signal.event) {
    case "secret_exfiltrated":
    case "cross_tenant_leakage":
    case "production_shell_executed":
    case "credential_use_after_revoke":
      return Severity.Critical;
    case "egress_with_secret_blocked":
    case "mcp_server_compromised":
    case "unsafe_tool_executed":
    case "egress_destination_out_of_policy":
    case "undeclared_tool_call":
    case "audit_gap":
    case "score_spike_after_network":
      return Severity.High;
    case "prompt_injection_detected":
    case "tool_denied_repeated":
    case "budget_exceeded":
    case "tool_retry_after_deny":
    case "eval_probe_suspected":
      return Severity.Medium;
    default:
      return Severity.Low;
  }
}

enum ContainmentAction {
  StopRun = "stop_run",
  DisableTool = "disable_tool",
  BlockEgress = "block_egress",
  RotateSecret = "rotate_secret",
  IsolateMemory = "isolate_memory",
  ReadOnlyMode = "read_only_mode",
  DisableMCP = "disable_mcp_server",
}

function buildContainmentPlan(signal: Signal): ContainmentAction[] {
  switch (signal.event) {
    case "secret_exfiltrated":
      return [
        ContainmentAction.StopRun,
        ContainmentAction.BlockEgress,
        ContainmentAction.RotateSecret,
        ContainmentAction.DisableTool,
      ];
    case "mcp_server_compromised":
      return [
        ContainmentAction.DisableMCP,
        ContainmentAction.BlockEgress,
        ContainmentAction.IsolateMemory,
      ];
    case "memory_poisoning_detected":
      return [ContainmentAction.StopRun, ContainmentAction.IsolateMemory];
    case "unsafe_tool_executed":
      return [
        ContainmentAction.StopRun,
        ContainmentAction.DisableTool,
        ContainmentAction.ReadOnlyMode,
      ];
    default:
      return [ContainmentAction.StopRun];
  }
}

/** Ordered IR path for containment / pivot (see §23 playbook). */
function autonomousContainmentSteps(): string[] {
  return [
    "stop_agent",
    "revoke_credentials",
    "block_egress",
    "save_tool_trace",
    "ignore_user_facing_summary",
    "pivot_check",
    "rotate_secrets",
    "notify_affected",
    "regression_eval",
  ];
}

function validateIncident(incident: Incident): void {
  if (!incident.id || !incident.title || !incident.owner) {
    throw new Error("incident has required empty fields");
  }

  if (!incident.severity || !incident.status) {
    throw new Error("incident has no severity or status");
  }

  if (incident.runIds.length === 0) {
    throw new Error("incident has no run ids");
  }

  if (!incident.detectedAt) {
    throw new Error("incident has no detected_at");
  }
}

function exportIncident(incident: Incident): string {
  validateIncident(incident);
  return JSON.stringify(incident, null, 2);
}
